OCPBUG 6952: Modified 'understanding how to override JVM max heap size' section (#68617)
openshift/openshift-docs: modules/nodes-cluster-resource-configure-jdk.adoc

. Overriding the JVM maximum heap size.

. Encouraging the JVM to release unused memory to the operating system, if appropriate.

. Ensuring all JVM processes within a container are appropriately configured.

Optimally tuning JVM workloads for running in a container is beyond the scope of this documentation, and may involve setting multiple additional JVM options.

[id="nodes-cluster-resource-configure-jdk-heap_{context}"]
== Understanding how to override the JVM maximum heap size

OpenJDK defaults to using a maximum of 25% of available memory (recognising any container memory limits in place) for "heap" memory. This default value is conservative, and in a properly-configured container environment, this value would result in 75% of the memory assigned to a container being mostly unused. A much higher percentage for the JVM to use for heap memory, such as 80%, is more suitable in a container context where memory limits are imposed on the container level.

Most of the Red Hat containers replace the OpenJDK default by means of a startup script - included in those containers - that set updated values when the JVM launches.

> For example, the Red Hat build of OpenJDK containers have a default value of 80%. This value can be set to a different percentage by defining the following environment variable:
> 
> [source,terminal]
> ----
> JAVA_MAX_RAM_RATIO
> ----

Reviewer: Does this need a block? it's not showing a command, as written
```suggestion
For example, the Red{nbsp}Hat build of OpenJDK containers have a default value of 80%. This value can be set to a different percentage by defining the `JAVA_MAX_RAM_RATIO` environment variable.
```

For other OpenJDK deployements, the default value of 25% can be changed using the following command:

.Example
[source,terminal]
----

> java -XX:MaxRAMPercentage=80.0

Reviewer: commands need a `$` prompt unless it's run as root (in which case it gets a `#`)
```suggestion
$ java -XX:MaxRAMPercentage=80.0
```

Java-based workloads can use the following JVM arguments to encourage the JVM to release unused memory to the operating system:

[source,terminal]
----
-XX:+UseParallelGC
-XX:MinHeapFreeRatio=5 -XX:MaxHeapFreeRatio=10 -XX:GCTimeRatio=4
-XX:AdaptiveSizePolicyWeight=90
----

These arguments are intended to return heap memory to the operating system whenever allocated memory exceeds 110% of in-use memory (`-XX:MaxHeapFreeRatio`), spending up to 20% of CPU time in the garbage collector (`-XX:GCTimeRatio`). At no time will the application heap allocation be less than the initial heap allocation (overridden by `-XX:InitialHeapSize` /
`-Xms`). Detailed additional information is available in link:https://developers.redhat.com/blog/2014/07/15/dude-wheres-my-paas-memory-tuning-javas-footprint-in-openshift-part-1/[Tuning Java's footprint in OpenShift (Part 1)], link:https://developers.redhat.com/blog/2014/07/22/dude-wheres-my-paas-memory-tuning-javas-footprint-in-openshift-part-2/[Tuning Java's footprint in OpenShift (Part 2)], and at link:https://developers.redhat.com/blog/2017/04/04/openjdk-and-containers/[OpenJDK

> and Containers].

Reviewer: Probably out of scope for this PR but it would be nice to organize this as a list of arguments and what they do instead of a block that lists them and then a paragraph under - it's a bit hard to follow

> and Containers].

Reviewer: Unwrap this line too?

[id="nodes-cluster-resource-configure-jdk-proc_{context}"]
== Understanding how to ensure all JVM processes within a container are appropriately configured

In the case that multiple JVMs run in the same container, it is essential to ensure that they are all configured appropriately. For many workloads it will be necessary to grant each JVM a percentage memory budget, leaving a perhaps substantial additional safety margin.

Many Java tools use different environment variables (`JAVA_OPTS`, `GRADLE_OPTS`, and so on) to configure their JVMs and it can be challenging to ensure that the right settings are being passed to the right JVM.

> The `JAVA_TOOL_OPTIONS` environment variable is always respected by the OpenJDK, and values specified in `JAVA_TOOL_OPTIONS` will be overridden by other options specified on the JVM command line. By default, to ensure that these options are used by default for all JVM workloads run in the Java-based agent image, the {product-title} Jenkins Maven agent image sets:

Reviewer: ```suggestion
The `JAVA_TOOL_OPTIONS` environment variable is always respected by the OpenJDK, and values specified in `JAVA_TOOL_OPTIONS` will be overridden by other options specified on the JVM command line. By default, to ensure that these options are used by default for all JVM workloads run in the Java-based agent image, the {product-title} Jenkins Maven agent image sets the following variable:
```

[id="nodes-cluster-resource-configure-jdk-heap_{context}"]
== Understanding how to override the JVM maximum heap size

OpenJDK defaults to using a maximum of 25% of available memory (recognising any container memory limits in place) for "heap" memory. This default value is conservative, and in a properly-configured container environment, this value would result in 75% of the memory assigned to a container being mostly unused. A much higher percentage for the JVM to use for heap memory, such as 80%, is more suitable in a container context where memory limits are imposed on the container level.

Most of the Red Hat containers replace the OpenJDK default by means of a startup script - included in those containers - that set updated values when the JVM launches.

For example, the Red Hat build of OpenJDK containers have a default value of 80%. This value can be set to a different percentage by defining the following environment variable:

[source,terminal]
----
JAVA_MAX_RAM_RATIO
----

For other OpenJDK deployements, the default value of 25% can be changed using the following command:

.Example
[source,terminal]
----

> java -XX:MaxRAMPercentage=80.0.

Reviewer 2: No trailing `.` please.

```suggestion
java -XX:MaxRAMPercentage=80.0
```

[id="nodes-cluster-resource-configure-jdk-heap_{context}"]
== Understanding how to override the JVM maximum heap size

> The default OpenJDK settings uses a maximum of 25% of the available memory for `heap` memory. This default is conservative, and it is in a `properly-configured` container environment. The default value would result in 75% of the memory assigned to an unused container. Higher default value such as 80% is more suitable for the container. 

Reviewer 2: The sentence: `This default is conservative, and it is in a properly-configured container environment.` doesn't make sense.

```suggestion
OpenJDK defaults to using a maximum of 25% of available memory (recognising
any container memory limits in place) for "heap" memory. This default value
is conservative, and in a properly-configured container environment, this value
would result in 75% of the memory assigned to a container being mostly unused.
A much higher limit, such as 80%, is more suitable.
```

They way you've rephrased it isn't meaning preserving, unfortunately.

Author: Let me rephrase it.

> Most of the {product-title} containers replaces the OpenJDK feature for a ctar-up script with OpenJDk default. 

Reviewer 2: It's not clear what this sentence means. What's a `ctar-up scrip`?

Typo: `OpenJDK`, not `OpenJDk`

Author: Ack.

> For example, the {product-title} build of OpenJDK containers default value to 80% can be turned to a different maximum by defining the following environment variable:

Reviewer 2: I don't think the use of `{product-title}` is correct here. There is no `OpenShift Container Platform build of OpenJDK`. The product is called `Red Hat build of OpenJDK`.

```suggestion
For example, the Red Hat build of OpenJDK containers have a default value of 80%. This value can be set to a different percentage by defining the following environment variable:
```

Author: Ack.

[source,terminal]
----

> JAVA_MAX_RAM_Ratio

Reviewer 2: Typo: `JAVA_MAX_MEM_RATIO`

Author: Ack.

----

> For other OpneJDK deployements, the default value of 25% can be changed using the following command:

Reviewer 2: Typo `OpenJDK`

[id="nodes-cluster-resource-configure-jdk-heap_{context}"]
== Understanding how to override the JVM maximum heap size

OpenJDK defaults to using a maximum of 25% of available memory (recognising any container memory limits in place) for "heap" memory. This default value is conservative, and in a properly-configured container environment, this value

> would result in 75% of the memory assigned to a container being mostly unused. A much higher limit, such as 80%, is more suitable.

Reviewer 2: ```suggestion
would result in 75% of the memory assigned to a container being mostly unused. A much higher percentage for the JVM to use for heap memory, such as 80%, is more suitable in a container context where memory limits are imposed on the container level.
```

> Most of the Red Hat containers replaces the OpenJDK feature for a start-up script with OpenJDK default. 

Reviewer 2: ```suggestion
Most of the Red Hat containers replace the OpenJDK default by means of a startup script - included in those containers - that set updated values when the JVM launches.
```

For example, the Red Hat build of OpenJDK containers have a default value of 80%. This value can be set to a different percentage by defining the following environment variable:

[source,terminal]
----
JAVA_MAX_RAM_RATIO
----

For other OpenJDK deployements, the default value of 25% can be changed using the following command:

.Example
[source,terminal]
----
java -XX:MaxRAMPercentage=80.0.
----

[id="nodes-cluster-resource-configure-jdk-unused_{context}"]
== Understanding how to encourage the JVM to release unused memory to the operating system

> By default, the OpenJDK does not aggressively return unused memory to the operating system. This may be appropriate for many containerized Java workloads, but notable exceptions include workloads where additional active processes co-exist with a JVM within a container, whether those additional processes are native, additional JVMs, or a combination of the two.

Reviewer 2: ```suggestion
By default, OpenJDK does not aggressively return unused memory to the operating system. This may be appropriate for many containerized Java workloads, but notable exceptions include workloads where additional active processes co-exist with a JVM within a container, whether those additional processes are native, additional JVMs, or a combination of the two.
```

> Java-based agents can use the following JVM arguments to encourage the JVM to release unused memory to the operating system:

Reviewer 2: ```suggestion
Java-based workloads can use the following JVM arguments to encourage the JVM to release unused memory to the operating system:
```

== Understanding how to override the JVM maximum heap size

For many Java workloads, the JVM heap is the largest single consumer of memory.

> Currently, the OpenJDK defaults to allowing up to 1/4 (1/`-XX:MaxRamPercentage`)

Reviewer 3: This statement is incorrect, for "bare metal/ VM", the default is 1/4th, for "container (cgroups v1 and v2)", it should be like: ubi8: 50% (set via Xmx) ubi9: 80% (set via -XX:MaxRAMPercentage)* (refer to https://access.redhat.com/solutions/4886081 )

Author: Acknowledged.

of the compute node's memory to be used for the heap, regardless of whether the
OpenJDK is running in a container or not. It is therefore *essential* to
override this behavior, especially if a container memory limit is also set.

> There are at least two ways the above can be achieved:

Reviewer 3: I think you can remove the entire paragraph, and describe it as https://access.redhat.com/solutions/4886081, refer to part "How the JVM calculates the heap and default values"

Author: Ack.

== Understanding how to override the JVM maximum heap size

For many Java workloads, the JVM heap is the largest single consumer of memory.

> Currently, on bare metal or VM setups, the OpenJDK's default heap value is 25%, 1/4th of the value (1/`-XX:MaxRamPercentage`). However, within containers using cgroups detection (v1 and v2 for JDK 11/17), the Xmx setting uses 50% of the maximum memory limits. In ubi9 - RHEL 9 images, the default allocation is 80%, leaving 20% for native use. It is therefore *essential* to

Reviewer 3: It should be: Currently, on bare metal or VM setups, the OpenJDK's default heap value is 25%, 1/4th of the compute node’s memory. However, within containers using cgroups detection (v1 and v2 for JDK 11/17), Xmx was set as 50% of the max memory resource limits by default. In ubi9 - RHEL 9 images, the default allocation is 80%(set via -XX:MaxRAMPercentage), leaving 20% for native use.

Author: Let me change it right away.

override this behavior, especially if a container memory limit is also set.

The `UseContainerSupport` flag is on by default in the JVM. It helps the JVM recognize Docker container operations, automatically fetching container-specific details like CPU count and total memory. Therefore, it is not necessary to manually define `JAVA_MAX_INITIAL_MEM`, as the JVM obtains maximum memory allocation from the pods directly.

> There are at least two ways the above can be achieved:

Reviewer 3: from https://access.redhat.com/solutions/4886081, MaxRAMFraction is deprecated(https://bugs.java.com/bugdatabase/view_bug.do?bug_id=JDK-8186248), so I think here the first way make no sense. And in the second way, you should add "MaxRAMPercentage" (override one of -XX:MaxRAM, -XX:MaxHeapSize , -Xmx or MaxRAMPercentage)

Author: Let me remove the first method.

Author: @lyman9966 For the second method, is there any link or something that can be added or any commands?

Reviewer 3: You can give an example similar to https://access.redhat.com/solutions/4886081 :
*Note: Prior to JDK17, MaxRAMPercentage is based on MaxRAM, not system memory. The default max value for MaxRAM is 128g. For example, this results in a 64g max heap, not 100g, in a container with a 200g memory limit: -XX:MaxRAMPercentage=50. A workaround for using MaxRAMPercentage in environments with >128g available memory is to explicitly set MaxRAM to a higher value. For example, this would result in a 100g max heap in a container with a 200g memory limit: -XX:MaxRAMPercentage=50 -XX:MaxRAM=200g). See [JDK-8222252](https://access.redhat.com/bounce/?externalURL=https%3A%2F%2Fbugs.openjdk.org%2Fbrowse%2FJDK-8222252).

Author: @lyman9966 How about adding 'IMPORTANT 'admonition here, to explain that note, something like this:-
"Prior to JDK17, note that `MaxRAMPercentage` used `MaxRAM`, instead of system memory. The default maximum value for `MaxRAM` is set to 128g. This means that in a container with a 200g memory limit, using `-XX:MaxRAMPercentage=50` will result in a 64g maximum heap, not 100g.

To address this issue in environments with more than 128g of available memory, a workaround is to explicitly set `MaxRAM` to a higher value. For instance, using `-XX:MaxRAMPercentage=50` `-XX:MaxRAM=200g` would result in a 100g maximum heap within a container with a 200g memory limit."

Reviewer 3: I'm OK for this change!

Author: Great. Let me move forward with further process.

* If the container memory limit is set and the default options are
   supported by the JVM, set `-XX:

>    -XX:+UseContainerSupport`.

Reviewer 3: line 41 and 42 should be: supported by the JVM, set `-XX:+UseContainerSupport`. (the original text has an additional -XX)

Author: Ack.

== Understanding how to override the JVM maximum heap size

For many Java workloads, the JVM heap is the largest single consumer of memory.

> Currently, on bare metal or VM setups, the OpenJDK's default heap value is 25%, 1/4th of the compute node’s memory. However, within containers using cgroups detection (v1 and v2 for JDK 11/17), Xmx was set as 50% of the max memory resource limits by default. In ubi9 - RHEL 9 images, the default allocation is 80%(set via -XX:MaxRAMPercentage), leaving 20% for native use. It is therefore essential to

Reviewer 4: THere's a lot of historical stuff here that I think can go. We now have full cgroups support in all JDK shipped by by RHEL (JDK8, 11 and newer).  I think it's important to clearly differentiate the behaviour of OpenJDK from the Red Hat OpenJDK containers, since  not all Java-using customers will be using the latter. How about this?

----

OpenJDK’s default heap value is 25% or <sup>1</sup>/<sub>4</sub><sup>th</sup> of
either the compute node’s memory or, if specified, the container memory limit.

The Red Hat UBI OpenJDK containers default to 80% of compute node memory or
container memory limit.

If this is not an appropriate value for your deployment, it can be overridden
with the command-line flag `-XX:MaxRAMPercentage`. For the Red Hat UBI OpenJDK
containers, this can be specified by setting the environment variable
`JAVA_MAX_MEM_RATIO` to the desired value e.g. `JAVA_MAX_MEM_RATIO=50`.

----

Additional context: 80% has been chosen as a sensible default for in-containers. If a user has specified a container memory limit, then it's a reasonable assumption that they want the JVM to use  virtually all of that memory, but we need to leave a little headroom for any non-java processes that might run in the same container (e.g. liveness probes and suchlike ,or things spawned by tools like `oc cp`). If the user has NOT set a memory limit, then we need to be careful that the JVM doesn't take all of the node's RAM, as this is very unlikely to be what they want. 80% is thought to be the right balance.

Reviewer 4: I'd actually delete everything else that follows in this file (except the header for the next section on the last line): notes about UseContainerSupport are historic, defining MaxRAM etc is not recommended if MaxRAMPercentage can achieve what is needed.

Reviewer 2: What Jon suggested sounds good to me!

Author: Okay..... Let me delete the note.

> override this behavior, especially if a container memory limit is also set.

Author: Ack.

[IMPORTANT]
====

> The `UseContainerSupport` flag is enabled by default in the JVM. It helps the JVM recognize Docker container operations by fetching the container details like CPU count and total memory. Therefore, it is not necessary to manually define `JAVA_MAX_INITIAL_MEM`, as the JVM obtains maximum memory allocation from the pods directly.

Author: Ack.

> The `UseContainerSupport` flag is enabled by default in the JVM. It helps the JVM recognize Docker container operations by fetching the container details like CPU count and total memory. Therefore, it is not necessary to manually define `JAVA_MAX_INITIAL_MEM`, as the JVM obtains maximum memory allocation from the pods directly.

Reviewer 4: `JAVA_MAX_INITIAL_MEM` is an environment variable tunable for the RHEL7 and RHEL8-based OpenJDK container images. It was deprecated in RHEL8 and removed for the RHEL9 images. It's not honoured by OpenJDK itself: it's handled by the run scripts in those containers. It is likely also honoured by the run scripts in various other product containers produced by Red Hat.

The run scripts translate it through to the JVM flag `-Xms`.

====

> To achieve this, directly override one of `-XX:MaxRAM`, `-XX:MaxHeapSize` or `MaxRAMPercentage`.

Author: Let me rephrase it.

== Understanding how to override the JVM maximum heap size

For many Java workloads, the JVM heap is the largest single consumer of memory.

> Currently, on bare metal or VM setups, the OpenJDK's default heap value is 25%, 1/4th of the compute node’s memory. However, within containers using cgroups detection (v1 and v2 for JDK 11/17), Xmx was set as 50% of the max memory resource limits by default. In ubi9 - RHEL 9 images, the default allocation is 80%(set via -XX:MaxRAMPercentage), leaving 20% for native use. It is therefore *essential* to

Author: Okay.

override this behavior, especially if a container memory limit is also set.
+
[IMPORTANT]
====

> The `UseContainerSupport` flag is on by default in the JVM. It helps the JVM recognize Docker container operations, automatically fetching container-specific details like CPU count and total memory. Therefore, it is not necessary to manually define `JAVA_MAX_INITIAL_MEM`, as the JVM obtains maximum memory allocation from the pods directly.

Author: Ack.

====

> To acheive this, directly override one of `-XX:MaxRAM`, `-XX:MaxHeapSize` or `MaxRAMPercentage`.

Author: Okay.

> To acheive this, directly override one of `-XX:MaxRAM`, `-XX:MaxHeapSize` or `MaxRAMPercentage`.

Author: Ack.

+

> This option involves hard-coding a value, but has the advantage of allowing a safety margin to be calculated.

Author: Ack.

+
[IMPORTANT]

> ----

Author: Ack.

> Prior to JDK17, note that `MaxRAMPercentage` used `MaxRAM`, instead of system memory. The default maximum value for `MaxRAM` is set to 128g. This means that in a container with a 200g memory limit, using `-XX:MaxRAMPercentage=50` will result in a 64g maximum heap, not 100g.

Author: It was a suggestion though. I will get it confirmed again. It might be better to remove the note.

> Prior to JDK17, note that `MaxRAMPercentage` used `MaxRAM`, instead of system memory. The default maximum value for `MaxRAM` is set to 128g. This means that in a container with a 200g memory limit, using `-XX:MaxRAMPercentage=50` will result in a 64g maximum heap, not 100g.

Author: Ack.

> To address this issue in environments with more than `128g` of available memory, a workaround is to explicitly set `MaxRAM` to a higher value. For instance, using `-XX:MaxRAMPercentage=50` `-XX:MaxRAM=200g` would result in a `100g` maximum heap within a container with a `200g` memory limit.

Reviewer 5: ```suggestion
To address this issue in environments with more than `128g` of available memory, set `MaxRAM` to a higher value. For instance, using `-XX:MaxRAMPercentage=50` and `-XX:MaxRAM=200g` would result in a `100g` maximum heap within a container with a `200g` memory limit.
```

Author: Addressed.

> To address this issue in environments with more than 128g of available memory, a workaround is to explicitly set `MaxRAM` to a higher value. For instance, using `-XX:MaxRAMPercentage=50` `-XX:MaxRAM=200g` would result in a 100g maximum heap within a container with a 200g memory limit.

Author: It is mentioned in the https://access.redhat.com/solutions/4886081, So the QE suggested it would be better to document it.

> To address this issue in environments with more than 128g of available memory, a workaround is to explicitly set `MaxRAM` to a higher value. For instance, using `-XX:MaxRAMPercentage=50` `-XX:MaxRAM=200g` would result in a 100g maximum heap within a container with a 200g memory limit.

Author: Ack.

> ----

Author: Ack.

and Containers].

[id="nodes-cluster-resource-configure-jdk-proc_{context}"]
== Understanding how to ensure all JVM processes within a container are appropriately configured

In the case that multiple JVMs run in the same container, it is essential to ensure that they are all configured appropriately. For many workloads it will be necessary to grant each JVM a percentage memory budget, leaving a perhaps substantial additional safety margin.

Many Java tools use different environment variables (`JAVA_OPTS`, `GRADLE_OPTS`, and so on) to configure their JVMs and it can be challenging to ensure that the right settings are being passed to the right JVM.

The `JAVA_TOOL_OPTIONS` environment variable is always respected by the OpenJDK, and values specified in `JAVA_TOOL_OPTIONS` will be overridden by other options specified on the JVM command line. By default, to ensure that these options are used by default for all JVM workloads run in the Java-based agent image, the {product-title} Jenkins Maven agent image sets:

[source,terminal]
----

> JAVA_TOOL_OPTIONS=" -Dsun.zip.disableMemoryMapping=true"

Reviewer 2: Drop the space.

```suggestion
JAVA_TOOL_OPTIONS="-Dsun.zip.disableMemoryMapping=true"
```
